src/pages/index.js

```diff
@@ -72,21 +89,105 @@ function enableValidationOnAllForms(config) {
+  }
+}
+
+async function initializeCards() {
+  try {
+    const initialCards = await siteApi.getInitialCards();
+    cardSection = new Section(
+      { items: initialCards, renderer: createCard },
+      ".cards__list"
+    );
+    cardSection.renderItems();
+  } catch (err) {
+    errorMessagePopup.setMessage(`${err}. Could not retrieve cards.`);
+    errorMessagePopup.open();
+  }
+}
+
+async function handleTrashButtonClick(card) {
+  try {
+    confirmCardDeletePopup.open();
+    confirmCardDeletePopup.setSubmitHandler(async () => {
+      await siteApi.deleteCard(card.id);
+      card.deleteCard();
+    });
+  } catch (err) {
+    errorMessagePopup.setMessage(`${err}. Could not delete card.`);
+    errorMessagePopup.open();
+  }
+}
+
+async function handleLikeButtonClick(card, likeButtonStateActive) {
+  try {
+    if (likeButtonStateActive) {
+      siteApi.unlikeCard(card.id);
+    } else {
+      siteApi.likeCard(card.id);
+    }
+  } catch (err) {
+    errorMessagePopup.setMessage(`${err}. Could not like/unlike card.`);
+    errorMessagePopup.open();
+  }
+}
+
 /* -------------------------------------------------------------------------- */
 /*                               Event Handlers                               */
 /* -------------------------------------------------------------------------- */
 
 //handle form submits. Class interactions handled through loose coupling
-function handleProfileEditSubmit({ title, description }) {
-  profileUserInfo.setUserInfo(title, description);
-  profileTitleTooltipHandler.handleTooltip();
-  profileDescriptionTooltipHandler.handleTooltip();
+async function handleProfileEditSubmit({ title, description }) {
+  try {
+    const fetchRequest = siteApi.updateUserInfo(title, description);
+    await profileEditPopup.applySavingAnimation(fetchRequest);
+    profileUserInfo.setUserInfo({
+      userName: title,
+      userDescription: description,
+    });
+    profileTitleTooltipHandler.handleTooltip();
+    profileDescriptionTooltipHandler.handleTooltip();
+  } catch (err) {
+    errorMessagePopup.setMessage(`${err}. Error setting user info.`);
+    errorMessagePopup.open();
+  }
 }
 
-function handleAddCardSubmit({ title, imageLink }) {
-  cardSection.addItem({ name: title, link: imageLink }, "prepend");
-  cardTooltipHandler.handleTooltip();
-  formValidators["modal-add-card-form"].resetFormValidation(true);
+async function handleAddCardSubmit({ title, imageLink }) {
+  try {
+    const cardData = { name: title, link: imageLink };
+    const fetchRequest = siteApi.addCard(cardData);
+    const card = await addCardPopup.applySavingAnimation(fetchRequest);
+    cardSection.addItem(card, "prepend");
+    cardTooltipHandler.handleTooltip();
+    formValidators["modal-add-card-form"].resetFormValidation(true);
+  } catch (err) {
+    errorMessagePopup.setMessage(`${err}. Could not add card.`);
+    errorMessagePopup.open();
+  }
+}
+
+async function handleProfilePictureSubmit(inputValues) {
+  try {
```

Comment:
**Could be improved**

If you did requests in a classic way with `renderLoading` then:

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```